## Summary for Part 2:
Owner:  Leana

#### Task:  
1.  Generate an update of the estimated number of OY in South King County.  
2.  In addition to the estimate, be sure to include a breakdown of the count of OY by Public Use Microdata Area (PUMA) within South King County:

**Note:  All numbers in this doc need to be scaled to show that they represent 5% of the population (check this)**

In [2]:
# Imports:
import psycopg2
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Establish DB:
DBNAME = "opportunity_youth"

# Create a connection to db
conn = psycopg2.connect(dbname=DBNAME)

**(a) Isolate South King County:**

In [14]:
# Find WA puma codes:
puma_names_2010_wa = pd.read_sql("SELECT * FROM puma_names_2010 WHERE state_name = 'Washington';", conn)
puma_names_2010_wa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   state_fips  56 non-null     object
 1   state_name  56 non-null     object
 2   cpuma0010   56 non-null     object
 3   puma        56 non-null     object
 4   geoid       56 non-null     object
 5   gisjoin     56 non-null     object
 6   puma_name   56 non-null     object
dtypes: object(7)
memory usage: 3.2+ KB


In [15]:
#  Show table
puma_names_2010_wa

state_fips                                         state_name cpuma0010  \
0          53  Washington                                    ...      1030   
1          53  Washington                                    ...      1031   
2          53  Washington                                    ...      1032   
3          53  Washington                                    ...      1032   
4          53  Washington                                    ...      1032   
5          53  Washington                                    ...      1032   
6          53  Washington                                    ...      1032   
7          53  Washington                                    ...      1032   
8          53  Washington                                    ...      1032   
9          53  Washington                                    ...      1033   
10         53  Washington                                    ...      1033   
11         53  Washington                                    ...      1033   
12         53  Washington                                    ...      1033   
13         53  Washington                                    ...      1034   
14         53  Washington                                    ...      1034   
15         53  Washington                                    ...      1035   
16         53  Washington                                    ...      1035   
17         53  Washington                                    ...      1035   
18         53  Washington                                    ...      1035   
19         53  Washington                                    ...      1036   
20         53  Washington                                    ...      1036   
21         53  Washington                                    ...      1036   
22         53  Washington                                    ...      1036   
23         53  Washington                                    ...      1037   
24         53  Washington                                    ...      1037   
25         53  Washington                                    ...      1038   
26         53  Washington                                    ...      1038   
27         53  Washington                                    ...      1038   
28         53  Washington                                    ...      1038   
29         53  Washington                                    ...      1038   
30         53  Washington                                    ...      1038   
31         53  Washington                                    ...      1038   
32         53  Washington                                    ...      1039   
33         53  Washington                                    ...      1040   
34         53  Washington                                    ...      1041   
35         53  Washington                                    ...      1042   
36         53  Washington                                    ...      1043   
37         53  Washington                                    ...      1044   
38         53  Washington                                    ...      1044   
39         53  Washington                                    ...      1044   
40         53  Washington                                    ...      1044   
41         53  Washington                                    ...      1044   
42         53  Washington                                    ...      1044   
43         53  Washington                                    ...      1044   
44         53  Washington                                    ...      1044   
45         53  Washington                                    ...      1044   
46         53  Washington                                    ...      1045   
47         53  Washington                                    ...      1046   
48         53  Washington                                    ...      1047   
49         53  Washington                                    ...      1047   
50         53  Washington

In [21]:
# Subset just the king county entries
king_county = puma_names_2010_wa[puma_names_2010_wa['puma_name'].str.match('^King County*')== True]
king_county

state_fips                                         state_name cpuma0010  \
37         53  Washington                                    ...      1044   
38         53  Washington                                    ...      1044   
39         53  Washington                                    ...      1044   
40         53  Washington                                    ...      1044   
41         53  Washington                                    ...      1044   
42         53  Washington                                    ...      1044   
43         53  Washington                                    ...      1044   
44         53  Washington                                    ...      1044   
45         53  Washington                                    ...      1044   
46         53  Washington                                    ...      1045   
47         53  Washington                                    ...      1046   

     puma    geoid    gisjoin  \
37  11606  5311606  G53011606   
38  11607  5311607  G53011607   
39  11608  5311608  G53011608   
40  11609  5311609  G53011609   
41  11610  5311610  G53011610   
42  11613  5311613  G53011613   
43  11614  5311614  G53011614   
44  11615  5311615  G53011615   
45  11616  5311616  G53011616   
46  11611  5311611  G53011611   
47  11612  5311612  G53011612   

                                            puma_name  
37  King County (Northwest)--Shoreline, Kenmore & ...  
38  King County (Northwest)--Redmond, Kirkland Cit...  
39  King County (Northwest Central)--Greater Belle...  
40  King County (Central)--Sammamish, Issaquah, Me...  
41  King County (Central)--Renton City, Fairwood, ...  
42  King County (Southwest Central)--Kent City    ...  
43  King County (Southwest)--Auburn City & Lakelan...  
44  King County (Southeast)--Maple Valley, Covingt...  
45  King County (Northeast)--Snoqualmie City, Cott...  
46  King County (West Central)--Burien, SeaTac, Tu...  
47  King County (Far Southwest)--Federal Way, Des ...

In [17]:
# Subset king county df to get south king county df using puma codes from king county df:
skc_puma_codes = ['11613', '11614', '11615', '11612']
skc = king_county[king_county['puma'].isin(skc_puma_codes)]
skc

state_fips                                         state_name cpuma0010  \
42         53  Washington                                    ...      1044   
43         53  Washington                                    ...      1044   
44         53  Washington                                    ...      1044   
47         53  Washington                                    ...      1046   

     puma    geoid    gisjoin  \
42  11613  5311613  G53011613   
43  11614  5311614  G53011614   
44  11615  5311615  G53011615   
47  11612  5311612  G53011612   

                                            puma_name  
42  King County (Southwest Central)--Kent City    ...  
43  King County (Southwest)--Auburn City & Lakelan...  
44  King County (Southeast)--Maple Valley, Covingt...  
47  King County (Far Southwest)--Federal Way, Des ...

**(b) Load pums_17 data to query relevant column subset:**

Columns we need for this task (would could also reduce these columns down at this point while we're just trying to count # of oy):
- rt (Record type - H = Housing record or group quarters Unit.  P = Person record)
- serialno (Housing unit/GQ person serial number - unique)
- puma
- agep
- esr (Employment status recode - e.g. 3 = Unemployed, 6 = Not in labor force)
- sch (School enrollment - e.g. 1 = not attended in last 3 months, 2 = public, 3 = private)
- schg (Grade level attending)
- schl (Educational attainment)
- nwab (Temporary absence from work)
- nwav (Available for work)
- nwla (On layoff from work)
- nwlk (Looking for work)
- sex

So!

In [18]:
QUERY = """
SELECT rt, serialno, agep, sex, st, puma, sch, schg, schl, esr, nwab, nwav, nwla, nwlk
FROM pums_2017
WHERE agep BETWEEN 16 AND 24"""

relevant_ages_pums = pd.read_sql(QUERY, conn)

`relevant_ages_pums` contains data for the appropriate age group (16-24) in all of WA.

In [19]:
# View info on this data:
relevant_ages_pums.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38170 entries, 0 to 38169
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   rt        38170 non-null  object 
 1   serialno  38170 non-null  object 
 2   agep      38170 non-null  float64
 3   sex       38170 non-null  object 
 4   st        38170 non-null  object 
 5   puma      38170 non-null  object 
 6   sch       38170 non-null  object 
 7   schg      23018 non-null  object 
 8   schl      38170 non-null  object 
 9   esr       38170 non-null  object 
 10  nwab      38170 non-null  object 
 11  nwav      38170 non-null  object 
 12  nwla      38170 non-null  object 
 13  nwlk      38170 non-null  object 
dtypes: float64(1), object(13)
memory usage: 4.1+ MB


This is saying there are 38,170 peopled aged 16-24 in all of Washington... this doesn't seem right...

**(c) If we then isolate South King County entries as per the codes we found above:**

In [13]:
# Now isolate skc rows:
skc_pums = relevant_ages_pums[relevant_ages_pums['puma'].isin(skc_puma_codes)]
skc_pums

rt       serialno  agep sex  st   puma sch  schg schl esr nwab nwav  \
32     P  2013000007063  19.0   1  53  11612   2    15   18   6    2    5   
36     P  2013000008046  17.0   2  53  11613   2    13   13   6    2    5   
48     P  2013000011255  17.0   2  53  11614   2    11   12   6    2    5   
54     P  2013000012970  21.0   2  53  11612   3    15   18   6    2    5   
57     P  2013000013525  18.0   2  53  11613   2    14   15   6    2    5   
...   ..            ...   ...  ..  ..    ...  ..   ...  ...  ..  ...  ...   
37134  P  2017001386502  18.0   1  53  11613   1  None   11   6    3    5   
37684  P  2017001464049  21.0   2  53  11613   1  None   16   6    3    5   
37718  P  2017001470135  23.0   1  53  11613   1  None   14   6    3    5   
38050  P  2017001518359  18.0   1  53  11613   1  None   11   6    3    5   
38130  P  2017001530818  23.0   1  53  11613   1  None   21   1    2    3   

      nwla nwlk  
32       2    2  
36       2    2  
48       2    2  
54       2    2  
57       2    2  
...    ...  ...  
37134    3    3  
37684    3    3  
37718    3    3  
38050    3    3  
38130    2    2  

[2359 rows x 14 columns]

#### To summarise so far:  There are 2359 individuals aged between 16-24 in South King County.  
Now we need to isolate which of these are categorised as 'opportunity youth' based on their schooling and employment.  
First, let's make the column names more helpful:

In [22]:
rename_columns = {'rt': 'record_type', 'serialno': 'id', 'puma': 'puma', 'agep': 'age', 'sex': 'sex', 'esr': 'employment_status', 
                  'sch': 'school_enrollment', 'schg': 'grade_attend', 'schl': 'education_attained', 'nwab': 'absent_from_work', 
                  'nwav': 'avail_for_work', 'nwla': 'layoff', 'nwlk': 'look_for_work'}
skc_pums.rename(columns = rename_columns, inplace = True)

In [23]:
# df of all individuals aged between 16-24 in South King County with updated column names
skc_pums

record_type             id   age sex  st   puma school_enrollment  \
32              P  2013000007063  19.0   1  53  11612                 2   
36              P  2013000008046  17.0   2  53  11613                 2   
48              P  2013000011255  17.0   2  53  11614                 2   
54              P  2013000012970  21.0   2  53  11612                 3   
57              P  2013000013525  18.0   2  53  11613                 2   
...           ...            ...   ...  ..  ..    ...               ...   
37134           P  2017001386502  18.0   1  53  11613                 1   
37684           P  2017001464049  21.0   2  53  11613                 1   
37718           P  2017001470135  23.0   1  53  11613                 1   
38050           P  2017001518359  18.0   1  53  11613                 1   
38130           P  2017001530818  23.0   1  53  11613                 1   

      grade_attend education_attained employment_status absent_from_work  \
32              15                 18                 6                2   
36              13                 13                 6                2   
48              11                 12                 6                2   
54              15                 18                 6                2   
57              14                 15                 6                2   
...            ...                ...               ...              ...   
37134         None                 11                 6                3   
37684         None                 16                 6                3   
37718         None                 14                 6                3   
38050         None                 11                 6                3   
38130         None                 21                 1                2   

      avail_for_work layoff look_for_work  
32                 5      2             2  
36                 5      2             2  
48                 5      2             2  
54                 5      2             2  
57                 5      2             2  
...              ...    ...           ...  
37134              5      3             3  
37684              5      3             3  
37718              5      3             3  
38050              5      3             3  
38130              3      2             2  

[2359 rows x 14 columns]

**(d) Isolate OY population by subsetting down to the youth who are unemployed AND not in school:**

In [24]:
# Codes relevant to unemployment and out of school:
oy_employment_status = ['3', '6']
oy_school_enrollment = ['1']
oy_grade_attend = ['bb']

In [25]:
# Create dataframe of just oy in skc:
oy_skc_pums = skc_pums[skc_pums['employment_status'].isin(oy_employment_status) & skc_pums['school_enrollment'].isin(oy_school_enrollment)]
oy_skc_pums.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 270 entries, 225 to 38050
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   record_type         270 non-null    object 
 1   id                  270 non-null    object 
 2   age                 270 non-null    float64
 3   sex                 270 non-null    object 
 4   st                  270 non-null    object 
 5   puma                270 non-null    object 
 6   school_enrollment   270 non-null    object 
 7   grade_attend        0 non-null      object 
 8   education_attained  270 non-null    object 
 9   employment_status   270 non-null    object 
 10  absent_from_work    270 non-null    object 
 11  avail_for_work      270 non-null    object 
 12  layoff              270 non-null    object 
 13  look_for_work       270 non-null    object 
dtypes: float64(1), object(13)
memory usage: 31.6+ KB


In [27]:
# Since grade_attend column is all null, we'll drop this column
oy_skc_pums.drop(columns='grade_attend', inplace = True)

KeyError: "['grade_attend'] not found in axis"

In [28]:
# See grade_attend dropped:
oy_skc_pums.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 270 entries, 225 to 38050
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   record_type         270 non-null    object 
 1   id                  270 non-null    object 
 2   age                 270 non-null    float64
 3   sex                 270 non-null    object 
 4   st                  270 non-null    object 
 5   puma                270 non-null    object 
 6   school_enrollment   270 non-null    object 
 7   education_attained  270 non-null    object 
 8   employment_status   270 non-null    object 
 9   absent_from_work    270 non-null    object 
 10  avail_for_work      270 non-null    object 
 11  layoff              270 non-null    object 
 12  look_for_work       270 non-null    object 
dtypes: float64(1), object(12)
memory usage: 29.5+ KB


In [29]:
# Preview our new dataframe of South King County OY:
oy_skc_pums.head()

record_type             id   age sex  st   puma school_enrollment  \
225           P  2013000047506  20.0   2  53  11615                 1   
234           P  2013000048962  22.0   2  53  11612                 1   
288           P  2013000058010  17.0   2  53  11614                 1   
327           P  2013000067168  24.0   2  53  11615                 1   
354           P  2013000070516  19.0   2  53  11612                 1   

    education_attained employment_status absent_from_work avail_for_work  \
225                 16                 6                2              5   
234                 19                 6                2              5   
288                 13                 6                2              5   
327                 18                 3                3              5   
354                 16                 6                2              5   

    layoff look_for_work  
225      2             2  
234      2             2  
288      2             2  
327      3             3  
354      2             2

**(e) Count the TOTAL number of OY from the last table:**

From `df.info()` we can see that we have 270 entries - this means the most updated estimate of the number of Opportunity Youth in South King County is 270 people (remember 270=5% of pop.).  

In [30]:
oy_skc_pums.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 270 entries, 225 to 38050
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   record_type         270 non-null    object 
 1   id                  270 non-null    object 
 2   age                 270 non-null    float64
 3   sex                 270 non-null    object 
 4   st                  270 non-null    object 
 5   puma                270 non-null    object 
 6   school_enrollment   270 non-null    object 
 7   education_attained  270 non-null    object 
 8   employment_status   270 non-null    object 
 9   absent_from_work    270 non-null    object 
 10  avail_for_work      270 non-null    object 
 11  layoff              270 non-null    object 
 12  look_for_work       270 non-null    object 
dtypes: float64(1), object(12)
memory usage: 29.5+ KB


#### Let's break this count down into individual ages:

In [31]:
oy_age_breakdown = oy_skc_pums.groupby(['age']).count()
oy_age_breakdown

record_type  id  sex  st  puma  school_enrollment  education_attained  \
age                                                                           
16.0           13  13   13  13    13                 13                  13   
17.0            9   9    9   9     9                  9                   9   
18.0           20  20   20  20    20                 20                  20   
19.0           36  36   36  36    36                 36                  36   
20.0           38  38   38  38    38                 38                  38   
21.0           28  28   28  28    28                 28                  28   
22.0           44  44   44  44    44                 44                  44   
23.0           41  41   41  41    41                 41                  41   
24.0           41  41   41  41    41                 41                  41   

      employment_status  absent_from_work  avail_for_work  layoff  \
age                                                                 
16.0                 13                13              13      13   
17.0                  9                 9               9       9   
18.0                 20                20              20      20   
19.0                 36                36              36      36   
20.0                 38                38              38      38   
21.0                 28                28              28      28   
22.0                 44                44              44      44   
23.0                 41                41              41      41   
24.0                 41                41              41      41   

      look_for_work  
age                  
16.0             13  
17.0              9  
18.0             20  
19.0             36  
20.0             38  
21.0             28  
22.0             44  
23.0             41  
24.0             41

In [32]:
# We can plot the above data - bar for each age.  x-axis = age, y-axis = population/count

**(f) Add a scaled column to show the 'true' population count of OY:**

In [33]:
oy_age_breakdown['scaled'] = oy_age_breakdown['id'] / 0.05

In [34]:
oy_age_breakdown

record_type  id  sex  st  puma  school_enrollment  education_attained  \
age                                                                           
16.0           13  13   13  13    13                 13                  13   
17.0            9   9    9   9     9                  9                   9   
18.0           20  20   20  20    20                 20                  20   
19.0           36  36   36  36    36                 36                  36   
20.0           38  38   38  38    38                 38                  38   
21.0           28  28   28  28    28                 28                  28   
22.0           44  44   44  44    44                 44                  44   
23.0           41  41   41  41    41                 41                  41   
24.0           41  41   41  41    41                 41                  41   

      employment_status  absent_from_work  avail_for_work  layoff  \
age                                                                 
16.0                 13                13              13      13   
17.0                  9                 9               9       9   
18.0                 20                20              20      20   
19.0                 36                36              36      36   
20.0                 38                38              38      38   
21.0                 28                28              28      28   
22.0                 44                44              44      44   
23.0                 41                41              41      41   
24.0                 41                41              41      41   

      look_for_work  scaled  
age                          
16.0             13   260.0  
17.0              9   180.0  
18.0             20   400.0  
19.0             36   720.0  
20.0             38   760.0  
21.0             28   560.0  
22.0             44   880.0  
23.0             41   820.0  
24.0             41   820.0

In [35]:
# Since all columns are the same count, we'll drop all the other columns:
oy_age_breakdown.drop(columns=['record_type', 'sex', 'st', 'puma', 
                               'school_enrollment', 'education_attained', 'employment_status', 
                               'absent_from_work', 'avail_for_work', 'layoff', 'look_for_work'], inplace = True)
oy_age_breakdown

id  scaled
age             
16.0  13   260.0
17.0   9   180.0
18.0  20   400.0
19.0  36   720.0
20.0  38   760.0
21.0  28   560.0
22.0  44   880.0
23.0  41   820.0
24.0  41   820.0

In [39]:
# From this, we can get the total oy-estimate, both scaled and 'true':
total_oy_est = oy_age_breakdown.sum()
total_oy_est

id         270.0
scaled    5400.0
dtype: float64

Further questioning:
- Do these numbers seem reasonable given that the estimated oy was 18,817 in 2016?  Does it seem reasonable that these numbers have reduced by 13,417?
- Next, create a new dataframe including the relevant columns necessary to answer questions/table cells for part 3.

Resources/References:

https://roadmapproject.org/wp-content/uploads/2018/09/Opportunity-Youth-2016-Data-Brief-v2.pdf

https://roadmapproject.org/wp-content/uploads/2018/09/Opportunity-Youth-Young-People-Disengaged-from-School-and-Work-in-South-King-County.pdf